### PR TITLE
Converted spraycharles to a CLI utility

### PR DESCRIPTION
**Pull request by puzzlepeaches**
# Description

This pull request converts spraycharles to a command line utility that can be accessed from anywhere on a users workstation. This is accomplished using the python poetry framework. Once installed the tool can be called using the spraycharles or sc commands while specifying the spray module. With poetry, support for pytest based testing scripts has also been added. A simple example test script has been included in this pull. 

All utility scripts (analyze, make_list, ntlm_challanger) are now callable as submodules of spraycharles. Note that these utilities are now no longer useable as standalone scripts and must be called from spraycharles directly. The JSON file list_elements.json has been moved the extras directory in the root of project. 

Output from the tool will be stored in a directory named `~/.spraycharles` created on first run. CSV file names are dynamically created in the format listed in the README.md. 

There may be other changes I am forgetting! This is going to be a big one. 

## Type of change

- [x] New feature (adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

To test the changes before committing to master, execute the following:

```
pipx install git+https://github.com/puzzlepeaches/spraycharles.git@refactor
```

We may need to pull this into a separate spraycharles branch in the main project for further testing. 

Broadly, the following tests were performed by me:

- [x] Tested all existing modules to the best of my ability following the changes.  
- [x] Tested submodules added to the framework

# Notes:

- Tried to add comments and docstrings for all new functionality.
- We will probably want to publish spraycharles to PyPi after everything has been tested. Going to leave this one to you as the owner of the project.
- Tried to update the CHANGELOG and versioning to reflect my changes in accordance with updates already merged to master since my last pull. 
- We may want to add another directory to ~/.spraycharles named files to allow users to easily store username/password lists and config files. The storage location for csv/log files was influenced by the behavior of manspider/crackmapexec.
- End users must use `spraycharles spray` to call the main functionality of the tool. 

# Coming soon
- Support for proxies 
- Module with the functionality of [Spray365](https://github.com/MarkoH17/Spray365) (This is going to be awesome!)
- Better handling for HTTP timeouts and other arbitrary connection errors. Been running into instances where the tool just raises an exception and closes without my knowledge. Going to try to integrate requests [retry](https://findwork.dev/blog/advanced-usage-python-requests-timeouts-retries-hooks/#retry-on-failure) functionality in addition to web hook notifications on fringe exceptions.
- Notifications when spray finishes completely. 
- Ability to call spraying modules like this: `spraycharles spray owa -u users.txt etc.`

Please let me know if I can be of any assistance during testing/integrating the merge. Have a great holiday!

### CODE REVIEW

**Comment by Tw1sm** [comment]
You're on a roll man! Changes sound sweet, hoping to review this week. My initial thought is that I'll merge this into a new branch and work on getting it published to PyPi, then merge into master, just to minimize the install changes users will go through.

Side note: might be good to add a `spraycharles list` command to just show the available target modules. Had a request for this the other day. I can worry about adding this after I merge into a new branch - mostly just wanted to write this down somewhere.

**Comment by puzzlepeaches** [comment]
Let me know if you need any help getting this published to PyPi. PyPi should support Github markdown now so we don't need to create a separate rst alternative. 

Will definitely work on getting that spraycharles list command implemented in some form. What do you think about a command structure like this?

```
#List valid modules
spraycharles spray list 

#Use specific module
spraycharles spray owa -u users.txt etc...
```

Want to make sure the final implementation dynamically resolves new modules when using the list subcommand so we don't have to update every time we add new spray targets. Let me know if you have any ideas on this. Planning to get this change pushed over by EOW. Thanks!

**Comment by Tw1sm** [comment]
I think `spraycharles spray list` or `spraycharles list` are both fine with me, I'd roll with whichever is easier to implement. 100% agree with dynamically resolving modules - since module class names have to match the file name, I'd say something does a directory list on `targets/` and grabs file names of `.py` files should work (excluding `__init__.py` I guess).

**Review by Tw1sm — COMMENTED** [review]
This is seriously great stuff 🔥. Add yourself as an author and I'll merge into the `pypi` branch. Looking into how to publish!

**Comment by puzzlepeaches** [comment]
Added the option to list modules in a rich table with:

```
sc list
```

iTerm221042022--06 01 22@2x


Still looking into how we convert the project to specify modules like:

```
sc spray owa -u users.txt etc...
```

Might not be part of this merge dependent on time constraints over in my neck of the woods. Will let you know.

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,53 @@
+[tool.poetry]
+name = "spraycharles"
+version = "1.06"
+description = "Low and slow password spraying tool, designed to spray on an interval over a long period of time."
+authors = [
+    "Matt Creel @Twism",
+    "Nicholas A @puzzlepeaches",
```
Comment (Tw1sm) [inline pyproject.toml:7]:
```suggestion
    "Matt Creel <@Twism>",
    "Nicholas A <@puzzlepeaches>",
```

**Comment by Tw1sm** [comment]
Added a quick fix for the author format, pipx threw an error when I tried to install cause it want the handles/email in `<>`. 

The `spraycharles list` command works great, gonna merge this in!

**Pull request merged**